mos-platform/pce/libpce/src/memory.S

```diff
@@ -0,0 +1,124 @@
+; Copyright (c) 2023 Adrian "asie" Siekierka
+;
+; Licensed under the Apache License, Version 2.0 with LLVM Exceptions,                                   
+; See https://github.com/llvm-mos/llvm-mos-sdk/blob/main/LICENSE for license                             
+; information.
+
+; This is a hack of the highest order.
+; The HuC6280 provides optimized (17+(6 cycles per byte)!) memory transfer
+; instructions, but they only take immediate arguments. Therefore, build
+; a subroutine in the imaginary register area and jump to it.
+
+.include "imag.inc"
+
+.global pce_memop
+.global pce_rmemcpy
+.global pce_memmove
+.global pce_memset
+
+    .section .text.pce_rmemcpy, "ax", @progbits
```

Comment:
For the little self-modifying assembly I've written in llvm-mos; I just wrote an assembly language routine, placed it in the data segment, and added labels to each instruction to be modified. Then, another routine can use those labels to adjust the routine, then simply jump to it. This minimizes the overhead of calling the routine, since only the portions that change between calls are overwritten, and the result is very readable, being actual assembly.

Reply:
Six of the eight bytes are caller-provided; as such, I believe the performance gain from doing so would be marginal (especially as the routine is laid out in such a way that - at least for `memcpy()` - two of the six caller-provided bytes are already in the right place at the moment of call), while the RAM use would increase by eight bytes (and we only have 8KB!).

The "ideal" way I'd like to resolve this is to emit the block copy instructions (TII/TDD) at the compilation stage if all the operands are known, minimizing the usage of these routines.